### DISCUSSION TITLE
Building my own element? (cohesive element)

### DISCUSSION THREAD

**Discussion by Traiwit** [idaholab/moose#18978] [Q&A Meshing] [answered]
Hi guys,

My team at work wants to investigate how to create our own element (cohesive element) by modifying pre-existing elements in moose.

This is for us to work around discontinuity in both tensor-mechanics and hydro problems, as we want our workflow to be as close as possible to what we do with Abaqus. ( I know MOOSE has the cohesive zone and the way to work with flow in fracture)

I suspect that we have to do that via Libmesh and then import it to MOOSE following this method: https://mooseframework.inl.gov/help/faq/faq_build_libmesh.html

Please give us some suggestions on where should we start to look into this.

Thank you very much.

Kind regards,
Traiwit

**Comment by GiudGiud** [comment, marked as answer]
Hi Traiwit

Depends, if you want to do this before the calculation you can use the Moose MeshGenerators to do this. 
https://mooseframework.inl.gov/syntax/Mesh/

If you want to do this during the calculation you can use the adaptivity system to do this. 
https://mooseframework.inl.gov/moose/syntax/Adaptivity/

I dont think we have this feature currently. Note that libmesh does not support arbitraty polyhedra so you will be restricted to the elements supported by libmesh. 
https://libmesh.github.io/doxygen/index.html

Guillaume

**Reply by hugary1995** [reply]
I think @Traiwit was asking about building his own element type, and I agree with him that seems to be libMesh-level development.

**Reply by Traiwit** [reply]
Thanks guys,

@hugary1995 so i should work on libmesh and just import it into MOOSE?

as I can see they don't have a cohesive element, but they have an element creation there

![image](https://user-images.githubusercontent.com/79562439/136018060-cb4934d4-e916-48c6-a930-89fe41680392.png)

https://libmesh.github.io/examples/introduction_ex1.html

**Reply by hugary1995** [reply]
I'm not a libMesh guy, @roystgnr is the expert. At the libMesh level, don't think of it as a cohesive element in the context of mechanics, it's just a lower dimensional face element connecting two neighboring volume eelements. @roystgnr can point you in the right direction of building such element.

**Reply by GiudGiud** [reply]
oh is a cohesive element a lower dimensional element between two elements or a merged element made of 2 elements? I thought the latter

**Reply by roystgnr** [reply]
Yeah, I'm a little confused there too.

For the former case: libMesh does support "multi-manifold" meshes, so e.g. you can have a sheet of 2D elements cutting through (or on the boundary of) a 3D mesh.  With Lagrange FE, you can connect the 2D elements to the neighboring elements' 3D nodes if you want their values to always match; with any FE you can have separate elements with their own (unique even if physically overlapping) nodes.

For the latter case: the only "macroelement" in libMesh is the Clough-Tocher triangle, and that had to be hand-coded (C1 continuity between subelements really constrains the function space there).  We don't have anything to let you create a larger element from a selection of smaller elements at runtime ... though I'd really like to add that someday; multiscale FEM is fantastic for complicated domains.

**Reply by Traiwit** [reply]
> oh is a cohesive element a lower-dimensional element between two elements or a merged element made of 2 elements? I thought the latter

I thought it's the first case, but not necessary has to be  'lower dimension, but it has ~0 thickness
![image](https://user-images.githubusercontent.com/79562439/136136659-eb8b3aec-1eb7-46be-860c-989696c15d95.png)
https://abaqus-docs.mit.edu/2017/English/SIMACAEELMRefMap/simaelm-c-cohesiveusage.htm

I see @arovinelli asked about this back in 2018, but I reckon Andrea has moved on and created his own version of the cohesive surface instead. https://groups.google.com/g/moose-users/c/hQkaOVbrm88

but yeah I would like to use the cohesive element in both hydro and mechanic (and maybe HM coupled) problems.

Kind regards,
Traiwit

**Reply by jiangwen84** [reply]
In my opinion, it is not trivial to add a cohesive element in libMesh and make it work seamlessly with MOOSE. That is the reason we originally implemented the DG-like CZM.  To large extend, the DG-like CZM is equivalent to the traditional CZM implementation in Abaqus. 

Maybe you could let us know why you want to implement traditional CZM? Are there any features that might be missing from the DG-like CZM? If so, we can probably add new features. That would be much easier to do than to reinvent the wheel.

**Reply by Traiwit** [reply]
> In my opinion, it is not trivial to add a cohesive element in libMesh and make it work seamlessly with MOOSE. That is the reason we originally implemented the DG-like CZM. To large extend, the DG-like CZM is equivalent to the traditional CZM implementation in Abaqus.
> 
> Maybe you could let us know why you want to implement traditional CZM? Are there any features that might be missing from the DG-like CZM? If so, we can probably add new features. That would be much easier to do than to reinvent the wheel.

Hi @jiangwen84 sorry for a late reply, I've been discussing with my team about this.

First of all, we are from Abaqus world where they have a cohesive element and this would be easier to implement to our existing complex meshes.

So we want to achieve a continuous solution field across several domains (domains are differently meshed, using tets and wedges). So there is no DG- or XFEM-approach needed. That would already be possible, however, due to the complexity of the discretization problem, we can only achieve those wedges to be flat [zero thickness] (the upper and lower triangle of the wedge share the same positions, but still have individual DOFs). 

The zero thickness element will also make moose run into problems due to the invalid jacobean. See two pics attached to illustrate this.

![image](https://user-images.githubusercontent.com/79562439/137260076-cf7ffc10-de32-4b62-b406-622a12d36418.png)

![image](https://user-images.githubusercontent.com/79562439/137260102-2646cf09-7a08-4f84-93a5-4f7c4e0ff714.png)

although we want to create an element that has zero thickness, we are not interested in modelling a discontinuity. We need to describe the behaviour of some material between adjacent tet-faces. This zone is in reality very thin compared to the surrounding material and the fact that we model it with zero-thickness in more or less a tweak to make it easier possible to introduce this zone at all.

the material we need to model needs its own state variable (damage, porosity etc) which means it must be able to be independently responding from the state of the adjacent surface due to its history.

#################################################################

we previously used wedges elements + NodalConstraints to achieve what we want for the hydro problems using MOOSE, but it would be nicer to have a cohesive element especially for mechanical (and HM) simulations

![image](https://user-images.githubusercontent.com/79562439/137260373-da5f25d5-653d-4d86-ad1e-72c738608a3c.png)
![image](https://user-images.githubusercontent.com/79562439/137260405-a01118ee-9b72-4ac5-b772-28ee6a49c955.png)
![image](https://user-images.githubusercontent.com/79562439/137260338-7016128c-7d34-425b-b0d2-70dad6b22316.png)

The question now would be, we should we look into in order to create a new element (cohesive element), is there any example? Our team previously have implemented FE element in Abaqus, so as long as we know where to start, we should be able to do it (ofc with your support 😃 )

Kind regards,
Traiwit

@GiudGiud @roystgnr @hugary1995 @arovinelli

**Reply by roystgnr** [reply]
So the idea here is that you have an infinitely thin element, e.g. a QUAD4 that's got two pairs of overlapping nodes, but you can't simply use an EDGE2 instead because you need 2 independent degrees of freedom for each variable on each node pair, but you also probably can't use a QUAD4 because the most basic FEM calculations are likely to break in a world where deta/dx or deta/dy is infinite?

I'd start by just positioning nodes to give you an infinitely thin QUAD4 and then see exactly what breaks.  libMesh does allow nodes to overlap, even from the same element, but I'm sure it'll end up screaming at you at some point when we try to do an inverse_map() and discover the map isn't invertible.  On the other hand, if you *don't* see any screaming (we don't use inverse_map() willy-nilly, since it's expensive, and we do have elements like pyramids that are usable despite not having an invertible map everywhere) then you might be able to do what you want just by creating your own quadrature rule?

**Reply by arovinelli** [reply]
@roystgnr there are specific quadrature rules for cohesive elements that allow nodes to be coincident. So what you describe is correct.

**Reply by Traiwit** [reply]
Hi guys, im not sure where exactly in MOOSE or libMesh should I start looking into this? is the file I'm looking for in here:

`~/anaconda3/pkgs/moose-libmesh-2021.09.15-build_0/libmesh/include`

@roystgnr

**Reply by roystgnr** [reply]
include/quadrature and src/quadrature for libMesh are the places to look when you're sure that all you need is a new quadrature rule.  I was hoping you'd be able to try out a mesh with singular elements first to make sure that the quadrature rule was your *only* obstacle, but I'm not sure what you could do to test that.  An H1 projection, but manually skip the un-integrable elements rather than trying to do a doomed FE reinit on them?  That'd be easy to do in a manual libMesh loop but I'm not sure how to do it in Moose.

**Reply by GiudGiud** [reply]
You cant develop in libmesh from a conda install of MOOSE.
You ll have to install libmesh manually, see this page: https://mooseframework.inl.gov/getting_started/installation/manual_installation_linux_lldb.html
Just use the LIBMESH_DIR environment variable to switch from one to another

**Reply by loganharbour** [reply]
The install description above is overkill. You can develop in libmesh by creating a separate conda env with moose-petsc instead of moose-libmesh.

**Reply by GiudGiud** [reply]
Yeah in the instructions above, please just follow the `install libmesh manually` part. so just
- unset LIBMESH_DIR
- running scripts/update_and_rebuild_libmesh.sh

**Comment by jiangwen84** [comment]
@Traiwit  I am not fully convinced that we have to use cohesive element for your problems. DG-interface is discontinuous, but with appropriate constraint, the field is continuous before damage.   You can define any stateful material properties, such as damage, porosity, on the interface which are independent of adjacent  surface. In addition, you can probably solve a lower-dimension problem, e.g. fluid, on the interface. 

In the meantime, I do not want to discourage you from exploring the cohesive element implementation in MOOSE. As @hugary1995 mentioned, creating element should be done in LibMesh and @roystgnr can provide you more insights. If my understanding is correct, the cohesive element can be standard quad in 2D with zero thickness, but the integration will be only performed in the middle line. Therefore, at minimum, the LibMesh needs to support zero thickness element type and new integration scheme for such element. @roystgnr is this hard to do? Thanks.
